docs/proposals/note-creation-filepath-fallback-chain.md

```diff
@@ -0,0 +1,72 @@
+# Templates v2 Proposal <!-- omit in TOC -->
+
+- [Introduction](#introduction)
+- [Proposal: "New note" path resolution](#proposal-new-note-path-resolution)
+  - [Phase 1: `<workspace>/.foam/templates/new-note.md`](#phase-1-workspacefoamtemplatesnew-notemd)
+  - [Phase 2: `<home_dir>/.foam/templates/new-note.md`](#phase-2-home_dirfoamtemplatesnew-notemd)
+  - [Phase 3: `foam.newNoteDirectory`](#phase-3-foamnewnotedirectory)
+  - [Phase 4: Match `SAME_AS_ACTIVE_NOTE` behaviour](#phase-4-match-same_as_active_note-behaviour)
+  - [Phase 5: Match `WORKSPACE_ROOT` behaviour](#phase-5-match-workspace_root-behaviour)
+  - [Phase 6: `<home_dir>`](#phase-6-home_dir)
+- [Progression of Foam usage maturity](#progression-of-foam-usage-maturity)
+
+## Introduction
+
+This document aims to standardize the fallback chain that is used by Foam when determining where to create a new note.
+Currently, each command has its own fallback chain, which leads to inconsistent behaviour in Foam.
+
+This document will guide the development direction that each note-creating-command will eventually be modified to follow.
+
+Note that throughout this doc, `foam.newNoteDirectory` is used as a placeholder to denote a new Foam setting that we would add, as discussed in [this comment](https://github.com/foambubble/foam/issues/670#issuecomment-860184377).
+
+## Proposal: "New note" path resolution
+
+### Phase 1: `<workspace>/.foam/templates/new-note.md`
+
+If `<workspace>/.foam/templates/new-note.md` exists, parse it looking for `filepath`:
+
+* An absolute path in `filepath` in `<workspace>/.foam/templates/new-note.md` should just use that absolute path.
+* A relative path in `filepath` in `<workspace>/.foam/templates/new-note.md` should be relative to the workspace.
+
+### Phase 2: `<home_dir>/.foam/templates/new-note.md`
+
+If `<workspace>/.foam/templates/new-note.md` does not exist (or no `filepath` attribute is defined in it), fall back to `<home_dir>/.foam/templates/new-note.md`
+
+* An absolute path in `filepath` in `<home_dir>/.foam/templates/new-note.md` should just use that absolute path.
+* A relative path in `filepath` in `<home_dir>/.foam/templates/new-note.md` should result in:
+  * **Option A**: Relative to the home directory
+    * This behaviour matches `<workspace>/.foam/templates/new-note.md` behaviour
+  * **Option B**: Relative to `foam.newNoteDirectory`
+    * I suspect is what you'd expect if you've already defined `foam.newNoteDirectory`
+  * **Option C**: Both. Relative to `foam.newNoteDirectory`, only when `foam.newNoteDirectory` is defined. Otherwise, relative to `<home_dir>`
```

Comment:
if you follow the `foam.newNoteDirectory` resolution from Phases 3, 4, 5, 6, we could at this level simply go with Option B, which would basically include Option C.

mmm.. not sure I am making sense.

Basically, a relative `filepath` would be matched against the resolved `foam.newNoteDirectory`.

`foam.newNoteDirectory` is resolved using the cases from Phase 3, 4, 5, 6 - and those are a superset of `Phase 2 -> Option C`. 

Let me know if this makes sense

Reply:
While Phase 6 is `<home_dir>`, it is really included just to handle the extreme edge case of having neither a file, nor a workspace open. If we resolve `foam.newNoteDirectory` from Phase 3-6, I expect that it would almost always get resolved by Phase 4, which changes the behaviour from what is currently described.

(i.e. Phase 4 "gets in the way", and is probably not the behaviour we want (creating a note relative to whatever you happen to have open)

---

I think you raise a good point to think about: Perhaps the resolution could be separated into two parts:

1. Resolving from the config files, falling back to `foam.newNoteDirectory`
2. Resolving `foam.newNoteDirectory`

I'll give it some more thought, but I'm not convinced yet.

---

### Potential changes to Phase 2 ("Option D")

I think the desired fallback behaviour of a relative path in Phase 2 might be:

Similar to being relative to Phases 3, 5, 6 (note: not Phase 4): 

* Similar to Phase 3: Relative to `foam.newNoteDirectory`
* Similar to Phase 5: Relative to Workspace
* Similar to Phase 6: Relative to `<home_dir>`

This would just affect Phase 2 with a relative path. In other cases, we'd we continue as described, doing Phase 3-6 (including 4)

Reply:
- In phase 3, how does a user set `SAME_AS_ACTIVE_NOTE` and `WORKSPACE_ROOT`? 
are those presented as "options" or "variables?

- an interesting things about phase 2 resolution is that you could have variables that only work in certain context (e.g. `WORKSPACE_ROOT`), I guess in that case they simply don't get expanded or, better, we can error out?

Reply:
@riccardoferretti 

I'm not proposing the addition of `SAME_AS_ACTIVE_NOTE` and `WORKSPACE_ROOT` into Foam, just equivalent behaviour as those.

I've reworded Phase 4 and 5 to be clearer in that regard.

Reply:
I think I understand, my question was about how the user could be deliberately set `foam.newNoteDirectory` to "same as active note" or "workspace root", is this a supported case?

Reply:
Hmmmm... 🤔 ... good question...

-----

`foam.newNoteDirectory` is defined as matching the spec used for `filepath` in the templates.

Once I convince VS Code to give extensions better snippet APIs, we'll be able to use [snippet variables](https://code.visualstudio.com/docs/editor/userdefinedsnippets#_variables) in these `filepath`s, including:

> * `TM_DIRECTORY`: The directory of the current document
> * `WORKSPACE_FOLDER`: The path of the opened workspace or folder

(Which work today in the rest of the snippets)

So at that point, users would be able to use those to get these behaviours.

* `SAME_AS_ACTIVE_NOTE` -> `"foam.newNoteDirectory": "$TM_DIRECTORY/$FOAM_TITLE"`
* `WORKSPACE_ROOT` -> `"foam.newNoteDirectory": "$WORKSPACE_FOLDER/$FOAM_TITLE"`